src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1106,83 +1104,151 @@ TargetEntryChangesValue(TargetEntry *targetEntry, Var *column, FromExpr *joinTre
 
 
 /*
- * RouterModifyTask builds a Task to represent a modification performed by
+ * RouterModifyJob builds a Job to represent a modification performed by
  * the provided query against the provided shard interval. This task contains
  * shard-extended deparsed SQL to be run during execution.
  */
-static Task *
-RouterModifyTask(Oid distributedTableId, Query *originalQuery,
-				 ShardInterval *shardInterval)
+static Job *
+RouterModifyJob(Query *originalQuery, Query *query, DeferredErrorMessage **planningError)
 {
-	Task *modifyTask = NULL;
-	DistTableCacheEntry *cacheEntry = DistributedTableCacheEntry(distributedTableId);
-
-	modifyTask = CitusMakeNode(Task);
-	modifyTask->jobId = INVALID_JOB_ID;
-	modifyTask->taskId = INVALID_TASK_ID;
-	modifyTask->taskType = MODIFY_TASK;
-	modifyTask->queryString = NULL;
-	modifyTask->anchorShardId = INVALID_SHARD_ID;
-	modifyTask->dependedTaskList = NIL;
-	modifyTask->replicationModel = cacheEntry->replicationModel;
+	Oid distributedTableId = ExtractFirstDistributedTableId(query);
+	List *taskList = NIL;
+	Job *job = NULL;
+	bool requiresMasterEvaluation = false;
+	bool deferredPruning = false;
 
-	if (originalQuery->onConflict != NULL)
+	if (!CanShardPrune(distributedTableId, query))
 	{
-		RangeTblEntry *rangeTableEntry = NULL;
+		/* there is a non-constant in the partition column, cannot prune yet */
+		taskList = NIL;
+		deferredPruning = true;
 
-		/* set the flag */
-		modifyTask->upsertQuery = true;
-
-		/* setting an alias simplifies deparsing of UPSERTs */
-		rangeTableEntry = linitial(originalQuery->rtable);
-		if (rangeTableEntry->alias == NULL)
+		/* must evaluate the non-constant in the partition column */
+		requiresMasterEvaluation = true;
+	}
+	else
+	{
+		taskList = RouterModifyTaskList(query, planningError);
+		if (*planningError)
 		{
-			Alias *alias = makeAlias(CITUS_TABLE_ALIAS, NIL);
-			rangeTableEntry->alias = alias;
+			return NULL;
 		}
+
+		/* determine whether there are function calls to evaluate */
+		requiresMasterEvaluation = RequiresMasterEvaluation(originalQuery);
 	}
 
-	if (shardInterval != NULL)
+	if (!requiresMasterEvaluation)
 	{
-		uint64 shardId = shardInterval->shardId;
-		StringInfo queryString = makeStringInfo();
+		/* no functions or parameters, build the query strings upfront */
+		RebuildQueryStrings(originalQuery, taskList);
+	}
+
+	job = CreateJob(originalQuery);
+	job->taskList = taskList;
+	job->requiresMasterEvaluation = requiresMasterEvaluation;
+	job->deferredPruning = deferredPruning;
+
+	return job;
+}
+
+
+/*
+ * CreateJob returns a new Job for the given query.
+ */
+static Job *
+CreateJob(Query *query)
+{
+	Job *job = NULL;
+
+	job = CitusMakeNode(Job);
```

Comment:
Ok, so one of the fields was omitted and one was duplicated, so yeah, this was a good suggestion.